File: Ballast.Core/src/Client/models/board.ts
import { ITile, Tile } from './tile';
import { ITileShape, TileShape } from './tile-shape';
import { IBoardType, BoardType } from './board-type';

export interface IBoard {
    boardTypeValue: number;
    id: string;
    tiles: ITile[];
    tileShapeValue: number;
}

export class Board implements IBoard {

    public readonly boardType: BoardType;
    public readonly boardTypeValue: number;
    public readonly id: string;
    public readonly tileShape: TileShape;
    public readonly tileShapeValue: number;
    public readonly tileMap: Map<string, Tile>;

    private cachedTiles?: Tile[];
    public get tiles() {
        if (!this.cachedTiles) {
            this.cachedTiles = Array.from(this.tileMap.values());
        }
        return this.cachedTiles;
    }

    public getTile(orderedTriple: number[]) {
        return this.tileMap.get(orderedTriple.toLocaleString());
    }

    private constructor(state: IBoard) {
        this.boardTypeValue = state.boardTypeValue;
        this.boardType = BoardType.fromValue(this.boardTypeValue);
        this.id = state.id;
        this.tileShapeValue = state.tileShapeValue;
        this.tileShape = TileShape.fromValue(state.tileShapeValue);
        this.tileMap = new Map(this.mapTiles(state.tiles));
    }

    public static fromObject(object: IBoard) {
        return new Board(object);
    }

    private *mapTiles(tiles: ITile[]) {
        let tileIterator = Tile.fromObjectList(tiles);
        for (let tile of tileIterator) {
            let item: [string, Tile] = [ tile.cubicCoordinates.toOrderedTriple().toLocaleString(), tile ];
            yield item;
        }
    }

}

File: Ballast.Core/src/Server/Models/Board.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ballast.Core.Models
{
    public class Board : IBoard
    {

        private readonly IEnumerable<Tile> _tiles;
        private readonly BoardType _boardType;
        private readonly TileShape _tileShape;

        public Guid Id { get; private set; }
        public IEnumerable<ITile> Tiles => _tiles;
        public IBoardType BoardType => _boardType;
        public ITileShape TileShape => _tileShape;

        private Board(IBoardType boardType, Guid id, IEnumerable<ITile> tiles, ITileShape tileShape)
        {
            _boardType = Models.BoardType.FromValue(boardType.Value);
            _tiles = tiles.Select(x => Tile.FromObject(x));
            _tileShape = Models.TileShape.FromValue(tileShape.Value);
            Id = id;
        }

        private Board(IBoard state) : this(
            boardType: state.BoardType,
            id: state.Id,
            tiles: state.Tiles,
            tileShape: state.TileShape
        ) { }
        
        public static Board FromObject(IBoard state) => new Board(state);

        public static Board FromProperties(IBoardType boardType, Guid id, IEnumerable<ITile> tiles, ITileShape tileShape) => new Board(
            boardType: boardType,
            id: id,
            tiles:tiles,
            tileShape: tileShape
        );

        public ICubicCoordinates GetRandomPassableCoordinates()
        {
            var passables = _tiles.Where(x => x.Terrain.Passable);
            var index = new Random().Next(1, passables.Count());
            return passables.ElementAt(index).CubicCoordinates;
        }

        public ITile GetTileFromCoordinates(ICubicCoordinates cubicCoordinates) =>
            _tiles.SingleOrDefault(x => x.CubicCoordinates.Equals(cubicCoordinates));

    }
}